### PR TITLE
Add user sign-in, add Firebase operation timestamp

### PR DESCRIPTION
**Pull request by rytrose**
This adds email/password user login in order to ensure people can't just refresh the page and get around our rules. It enables other features based on having persistent users, too.

It adds a DELETE admin endpoint that removes all Firebase users - **VERY DANGEROUS**, but was needed to delete the couple thousand anonymous users we had created.

I'll comment the key points in the diff.

### CODE REVIEW

File: server/client/js/main.js [hunk 1]
```diff
@@ -6,64 +6,158 @@ let users = db.collection('users');
...
+                        $("#signUp").show();
+                        break;
+                    case "auth/wrong-password":
+                        alert("Incorrect password.");
+                        break;
+                    default:
+                        alert(error.message);
+                }
+            });
+    });
```
Comment (rytrose) [inline server/client/js/main.js]: Sign-in function, allows for validation/UI logic on sign-in.

File: server/client/js/main.js [hunk 2]
```diff
@@ -6,64 +6,158 @@ let users = db.collection('users');
...
+                        alert("Invalid email.");
+                        break;
+                    case "auth/weak-password":
+                        alert("Password too weak.");
+                        break;
+                    default:
+                        alert(error.message);
+                }
+            });
+    });
```
Comment (rytrose) [inline server/client/js/main.js]: Sign-up logic, allows for validation/UI logic on sign-up.

File: server/client/js/main.js [hunk 3]
```diff
@@ -6,64 +6,158 @@ let users = db.collection('users');
...
+
+        auth.signOut();
+        $("#content").hide();
+        $("#signIn").show();
+    });
 });
 
 // Once logged in
-firebase.auth().onAuthStateChanged((user) => {
+firebase.auth().onAuthStateChanged(user => {
```
Comment (rytrose) [inline server/client/js/main.js:101]: This function is fired on both initial page load, and after sign-in/sign-up.

The main things to know are:
- If `user` exists, a user is signed in. That may be as a result of the sign-in flow/sign-up flow during this page view, or from Firbase maintaining the login across page open/close.
- If `user` does not exist, a user is not signed in and the sign-in/up UI should be shown.

File: server/client/js/main.js [hunk 4]
```diff
@@ -6,64 +6,158 @@ let users = db.collection('users');
...
+                    "userID": user.uid
+                })
+                    .then(res => {
+                        if(res.error) {
+                            // Client already connected
+                            alert("You're already logged in in another tab.");
+                            $("#content").remove();
+                        }
+                    })
+                    .catch(error => console.log("unable to announce:", error));
```
Comment (rytrose) [inline server/client/js/main.js:90]: The `announce` type message now has a response which returns an `error` property in the body should the user be signed-in and have the page open in another tab. I suggest we don't support this, but we could figure that out.

File: server/client/js/main.js [hunk 5]
```diff
@@ -102,16 +196,33 @@ rooms.where('active', '==', true).onSnapshot((snapshot) => {
                             $("#operations").append(`<p>${JSON.stringify(operation)}</p>`)
                         }
                         $("#operate").click(() => {
+                            console.log("click!");
+                            if (lastOperation) {
+                                if (Date.now() < lastOperation + roomData.submitTimeout) {
```
Comment (rytrose) [inline server/client/js/main.js:138]: Upon hitting submit, check if the user's last submit was long enough ago.

File: server/client/js/main.js [hunk 6]
```diff
@@ -102,16 +196,33 @@ rooms.where('active', '==', true).onSnapshot((snapshot) => {
...
+                            $("#operations").append(`<p>${JSON.stringify(operation)}</p>`);
                             socket.send({
                                 "type": "operation",
                                 "roomName": roomName,
                                 "operation": operation
                             });
+
+                            // Update last operation timestamp
+                            lastOperation = Date.now();
+                            userDoc.set({
```
Comment (rytrose) [inline server/client/js/main.js:157]: When submitting, store the submit timestamp for this user in Firebase.

File: server/client/js/main.js (hunk 1 above)
Comment (universalsequences) [reply server/client/js/main.js]: Does firebase force users to choose strong passwords?

Im a little worried of having too much friction to start messing around in the installation. 

How bad would it be to just ask people for a display name and that’s it? Would that make it too easy to make a bunch of fake names to bypass the rules?

File: server/client/js/main.js (hunk 4 above)
Comment (universalsequences) [reply server/client/js/main.js:90]: Do you mean we shouldnt support limiting users to one tab? Is it just that it adds complexity?

**Comment by universalsequences** [comment]
Sorry for taking so long to review! This looks all great, just wondering how necessary it is to have people choose passwords/use emails vs a “screen name”. 

With a screenname, you could have people take on identities, that we communicate across the room. Like when people do things, the UI could be like “So and só made this change here”

Also, does firebase force us to use a google style sign up/login, like when apps say “sign up with google?”. Or do we take care of all the login UI?

File: server/client/js/main.js (hunk 1 above)
Comment (rytrose) [reply server/client/js/main.js]: It does require some strength, but it's not much -- just needs to be 6 characters or more.

We could potentially ask for just a display name, would need a bit more work but it's doable if we want to go with that.

File: server/client/js/main.js (hunk 4 above)
Comment (rytrose) [reply server/client/js/main.js:90]: I suggest we _don't_ allow users to operate in more than one tab. Not sure how we would resolve conflicts and it doesn't make much sense to me.

**Comment by rytrose** [comment]
The problem with just a screenname is that you could impersonate someone easily! It doesn't require a real email (no validation or anything), but I do think a username of sorts and password is needed. Otherwise yes, you can circumvent the rules we want to enforce easily.

File: server/client/js/main.js (hunk 4 above)
Comment (universalsequences) [reply server/client/js/main.js:90]: Yeah I agree, that would just be them circumventing the rules + making it complicated for us

**Comment by universalsequences** [comment]
if we don’t allow people to use a screenname for that already exists, then they can’t impersonate. We could store a localstorage cookie (in the front end) with their username they choose. And check if that exists, otherwise ask them to choose a name. 

I guess someone could hack their chrome to fake that localstorage cookie, but I don’t think someone’s gonna go to that length. 

I don’t think it would be that much work, could just use what you have already here, and just submit the password as the username, everytime. And check for the cookie from the frontend, whether you’re allowed to use that username.

Does that make sense? I just feel like login with password makes it feel like an Application vs an installation. We should probably ask Avneesh for his thoughts

**Comment by Ryan-Rose-Bose** [comment]
True, yeah that can be done! I'll take a crack at that. I'd rather integrate it more formally with Firebase though, instead of manually managing the cookie. It's definitely possible though! If you don't oppose, I'll merge this and work off of this as a starting point.

**Comment by rytrose** [comment]
True, yeah that can be done! I'll take a crack at that. I'd rather integrate it more formally with Firebase though, instead of manually managing the cookie. It's definitely possible though! If you don't oppose, I'll merge this and work off of this as a starting point.

**Pull request closed without merging**